File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;
    
    public Node(int _val) {
        val = _val;
        next = null;
        random = null;
    }
}
*/

public class Solution {
    public Node copyRandomList(Node head) {
        if(head == null) return null;
        var set = new Dictionary<Node, Node>();
        var temp = new Node(-1);
        temp.next = head;        
        while(head != null)
        {
            var copyNode = new Node(head.val);
            set.Add(head, copyNode);
            head = head.next;
        }

        head = temp.next;        

        while(head != null)
        {            
            set[head].next = head.next == null ? null : set[head.next];
            set[head].random = head.random == null ? null : set[head.random];

            head = head.next;
        }

        return set[temp.next];
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []
        dummy = result = ListNode(-1)

        # add initial items to the heap
        for n in lists:
            heapq.heappush(heap, HeapNode(n))
        
        while heap:
            h_node = heapq.heappop(heap)
            result.next = h_node.node
            if h_node.node.next:
                heapq.heappush(heap, HeapNode(h_node.node.next))
            result = result.next
        

        return dummy.next

class HeapNode:
    def __init__(self, node: ListNode):
        self.node = node
    def __lt__(self, other: ListNode):
        return self.node.val < other.node.val
